source/opt/inst_bindless_check_pass.cpp

```diff
@@ -0,0 +1,262 @@
+static const int kInstBindlessOutError = 0;
+static const int kInstBindlessOutDescIndex = 1;
+static const int kInstBindlessOutDescBound = 2;
+static const int kInstBindlessOutRecordSize = 3;
+
+namespace spvtools {
+namespace opt {
+
+void InstBindlessCheckPass::GenBindlessCheckCode(
+    BasicBlock::iterator ref_inst_itr,
+    UptrVectorIterator<BasicBlock> ref_block_itr,
+    uint32_t function_idx,
+    uint32_t instruction_idx,
+    uint32_t stage_idx,
+  std::vector<std::unique_ptr<BasicBlock>>* new_blocks) {
+  // Look for reference through bindless descriptor. If not, return.
+  std::unique_ptr<BasicBlock> new_blk_ptr;
+  uint32_t imageId;
+  switch (ref_inst_itr->opcode()) {
+    case SpvOp::SpvOpImageSampleImplicitLod:
+    case SpvOp::SpvOpImageSampleExplicitLod:
+    case SpvOp::SpvOpImageSampleDrefImplicitLod:
+    case SpvOp::SpvOpImageSampleDrefExplicitLod:
+    case SpvOp::SpvOpImageSampleProjImplicitLod:
+    case SpvOp::SpvOpImageSampleProjExplicitLod:
+    case SpvOp::SpvOpImageSampleProjDrefImplicitLod:
+    case SpvOp::SpvOpImageSampleProjDrefExplicitLod:
+    case SpvOp::SpvOpImageGather:
+    case SpvOp::SpvOpImageDrefGather:
+    case SpvOp::SpvOpImageQueryLod:
+    case SpvOp::SpvOpImageSparseSampleImplicitLod:
+    case SpvOp::SpvOpImageSparseSampleExplicitLod:
+    case SpvOp::SpvOpImageSparseSampleDrefImplicitLod:
+    case SpvOp::SpvOpImageSparseSampleDrefExplicitLod:
+    case SpvOp::SpvOpImageSparseSampleProjImplicitLod:
+    case SpvOp::SpvOpImageSparseSampleProjExplicitLod:
+    case SpvOp::SpvOpImageSparseSampleProjDrefImplicitLod:
+    case SpvOp::SpvOpImageSparseSampleProjDrefExplicitLod:
+    case SpvOp::SpvOpImageSparseGather:
+    case SpvOp::SpvOpImageSparseDrefGather:
+    case SpvOp::SpvOpImageFetch:
+    case SpvOp::SpvOpImageRead:
+    case SpvOp::SpvOpImageQueryFormat:
+    case SpvOp::SpvOpImageQueryOrder:
+    case SpvOp::SpvOpImageQuerySizeLod:
+    case SpvOp::SpvOpImageQuerySize:
+    case SpvOp::SpvOpImageQueryLevels:
+    case SpvOp::SpvOpImageQuerySamples:
+    case SpvOp::SpvOpImageSparseFetch:
+    case SpvOp::SpvOpImageSparseRead:
+    case SpvOp::SpvOpImageWrite:
+      imageId =
+          ref_inst_itr->GetSingleWordInOperand(kSpvImageSampleImageIdInIdx);
+      break;
+    default:
+      return;
+  }
+  Instruction* imageInst = get_def_use_mgr()->GetDef(imageId);
+  uint32_t loadId;
+  Instruction* loadInst;
+  if (imageInst->opcode() == SpvOp::SpvOpSampledImage) {
+    loadId = imageInst->GetSingleWordInOperand(kSpvSampledImageImageIdInIdx);
+    loadInst = get_def_use_mgr()->GetDef(loadId);
+  }
+  else if (imageInst->opcode() == SpvOp::SpvOpImage) {
+    loadId = imageInst->GetSingleWordInOperand(kSpvImageSampledImageIdInIdx);
+    loadInst = get_def_use_mgr()->GetDef(loadId);
+  }
+  else {
+    loadId = imageId;
+    loadInst = imageInst;
+    imageId = 0;
+  }
+  if (loadInst->opcode() != SpvOp::SpvOpLoad) {
+    assert(false && "expected image load");
```

Comment:
I think this assert will get triggered.  Depending on the way the front-end generates code, you could go through `OpLoad->OpSampledImage->OpImage`.  We should try to handle all legal ways of manipulating images.

Reply:
While someone could create this sequence by hand, what I was thinking here is that there was no way this sequence could be generated directly either from GLSL or HLSL. Can you think of a counter example? One might also ask: is there any "reasonable" sequence of hand-written SPIR-V and spirv-opt passes that would result in this sequence? I guess I haven't completely convinced myself the answer is no, it seemed very unlikely so I didn't want to spend a lot of time on such patterns right now. In general, I am recommending that users run legalization/optimization passes first to improve the analysis of instrumentation. Instrumentation might reasonably expect that simplification would be run first and would optimize the result of this sequence to the load. For the moment I will replace the assert with a TODO to consider other possibilities.

Reply:
Okay